File: src/sandbox/containers/playground/cpp/timeline.cpp
#pragma once

#include "timeline.h"

using namespace std;
using namespace std::chrono;

TIMELINE::TIMELINE()
{
    m_constants["elapsedTime"] = 0.f;
    m_constants["elapsedTimeLevel"] = 0.f;
}

void TIMELINE::tick()
{
    if (!m_active) {
        return;
    }

    auto now = duration_cast< ms >(system_clock::now().time_since_epoch());
    auto dt = now - m_startTime;
    m_constants["elapsedTime"] = duration<float>(dt - m_elapsedTimeLevel).count();
    m_constants["elapsedTimeLevel"] = duration<float>(m_elapsedTimeLevel).count();
    m_elapsedTimeLevel = dt; 
}

bool TIMELINE::isStopped() const 
{
    return !m_active;
}

void TIMELINE::start()
{
    m_constants["elapsedTime"] = 0.f;
    m_constants["elapsedTimeLevel"] = 0.f;

    m_startTime = duration_cast< ms >(system_clock::now().time_since_epoch());
    m_elapsedTimeLevel = ms::zero();
    m_active = true;
    cout << "emitter started" << endl;
}

void TIMELINE::stop() 
{
    m_active = false;
    cout << "emitter stopped" << endl;
}

File: src/lib/fx/timeline.ts
import { verbose } from '@lib/common';

export function make() {
    let pauseTime: number;
    let pauseDelay: number;
    let startTime: number;
    let elapsedTimeLevel: number;
    let active: boolean;
    let paused: number;
    let frameNumber: number;

    const constants = {
        elapsedTime: 0,
        elapsedTimeLevel: 0,
        frameNumber: 0
    };

    function stop() {
        active = false;
        paused = 0;
        verbose('timeline stopped');
    }

    function start() {
        constants.elapsedTime = 0;
        constants.elapsedTimeLevel = 0;
        constants.frameNumber = 0;

        paused = 0;
        pauseDelay = 0;
        pauseTime = 0;
        frameNumber = 0;
        startTime = Date.now();
        elapsedTimeLevel = 0;
        active = true;
        verbose('timeline started');
    }

    function tick() {
        if (isStopped() || isPaused()) {
            return;
        }

        const dt = Date.now() - startTime - pauseDelay;
        constants.elapsedTime = (dt - elapsedTimeLevel) / 1000;
        constants.elapsedTimeLevel = elapsedTimeLevel / 1000;
        constants.frameNumber = ++frameNumber;
        elapsedTimeLevel = dt;
        // frameNumber++;
    }

    function isStopped() {
        return !active;
    }

    function isPaused() {
        return !!paused;
    }

    function pause()
    {
        if (isStopped())
            return;
        
        if (!isPaused()) {
            pauseTime = Date.now();
        }

        paused++;
    }

    function unpause()
    {
        if (isStopped() || !isPaused())
            return;
        
        paused--;

        if (!isPaused()) {
            pauseDelay += Date.now() - pauseTime;
        }
    }

    function getConstants()
    {
        return constants;
    }

    return {
        getConstants,
        start,
        stop,
        tick,
        isStopped,

        pause,
        unpause,
        isPaused
    };
}

export type ITimeline = ReturnType<typeof make>;
